zebra-state/src/service/finalized_state.rs

```diff
@@ -363,32 +363,12 @@ impl FinalizedState {
             let sapling_root = sapling_note_commitment_tree.root();
             let orchard_root = orchard_note_commitment_tree.root();
 
-            // Create the history tree if it's the Heartwood activation block.
-            let heartwood_height = NetworkUpgrade::Heartwood
-                .activation_height(self.network)
-                .expect("Heartwood height is known");
-            match height.cmp(&heartwood_height) {
-                std::cmp::Ordering::Less => assert!(
-                    history_tree.is_none(),
-                    "history tree must not exist pre-Heartwood"
-                ),
-                std::cmp::Ordering::Equal => {
-                    history_tree = Some(HistoryTree::from_block(
-                        self.network,
-                        block.clone(),
-                        &sapling_root,
-                        &orchard_root,
-                    )?);
-                }
-                std::cmp::Ordering::Greater => history_tree
-                    .as_mut()
-                    .expect("history tree must exist Heartwood-onward")
-                    .push(block.clone(), &sapling_root, &orchard_root)?,
-            }
+            history_tree.push(self.network, block.clone(), sapling_root, orchard_root)?;
```

Comment:
Style: Can we open a follow-up ticket to:
1. Move the note commitment tree and history tree updates outside the database batch?
2. Split out the chain order assertion, the calculations, and the write batch into separate methods?

If we move the calculations, the batch closure becomes infallible, because it only contains serialization and deserialization.

Splitting the method should make reviews a bit easier - this method is getting really long!
(Also we should be able to remove the closure entirely, and just return the `WriteBatch` from the method.)

Reply:
Created https://github.com/ZcashFoundation/zebra/issues/2588